presto-main/src/main/java/com/facebook/presto/operator/Driver.java

```diff
@@ -317,7 +317,7 @@ private void processNewSources()
                 }
 
                 // if current operator is finished...
-                if (current.isFinished()) {
+                if (current.isFinished() && !next.isFinished() && isBlocked(next).isDone()) {
```

Comment:
@dain the idea is that would simplify `Operator` implementation at no cost. Note that Driver won't call `Operator.addInput` when operator is blocked, so why not guarantee `Operator.finish()` won't be called when blocked either?